pkg/action/git-credential.go

```diff
@@ -0,0 +1,192 @@
+	Host     string
+	Path     string
+	Username string
+	Password string
+}
+
+// WriteTo writes the given credentials to the given io.Writer in the git-credential format
+func (c *gitCredentials) WriteTo(w io.Writer) (int64, error) {
+	var n int64
+	if c.Protocol != "" {
+		i, err := io.WriteString(w, "protocol="+c.Protocol+"\n")
+		n += int64(i)
+		if err != nil {
+			return n, err
+		}
+	}
+	if c.Host != "" {
+		i, err := io.WriteString(w, "host="+c.Host+"\n")
+		n += int64(i)
+		if err != nil {
+			return n, err
+		}
+	}
+	if c.Path != "" {
+		i, err := io.WriteString(w, "path="+c.Path+"\n")
+		n += int64(i)
+		if err != nil {
+			return n, err
+		}
+	}
+	if c.Username != "" {
+		i, err := io.WriteString(w, "username="+c.Username+"\n")
+		n += int64(i)
+		if err != nil {
+			return n, err
+		}
+	}
+	if c.Password != "" {
+		i, err := io.WriteString(w, "password="+c.Password+"\n")
+		n += int64(i)
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+func parseGitCredentials(r io.Reader) (*gitCredentials, error) {
+	rd := bufio.NewReader(r)
+	c := &gitCredentials{}
+	for {
+		key, err := rd.ReadString('=')
+		if err != nil {
+			if err == io.EOF {
+				return c, nil
+			}
+			return nil, err
+		}
+		key = strings.TrimSuffix(key, "=")
+		val, err := rd.ReadString('\n')
+		if err == io.EOF {
+			err = io.ErrUnexpectedEOF
+		}
+		if err != nil {
+			return nil, err
+		}
+		val = strings.TrimSuffix(val, "\n")
+		switch key {
+		case "protocol":
+			c.Protocol = val
+		case "host":
+			c.Host = val
+		case "path":
+			c.Path = val
+		case "username":
+			c.Username = val
+		case "password":
+			c.Password = val
+		}
+	}
+}
+
+// GitCredentialBefore is executed before another git-credential command
+func (s *Action) GitCredentialBefore(ctx context.Context, c *cli.Context) error {
+	if !ctxutil.IsStdin(ctx) {
+		return ExitError(ctx, ExitUsage, nil, "missing stdin from git")
+	}
+	return s.Initialized(ctx, c)
+}
+
+// GitCredentialGet returns a credential to git
+func (s *Action) GitCredentialGet(ctx context.Context, c *cli.Context) error {
+	ctx = sub.WithAutoSync(ctx, false)
+	cred, err := parseGitCredentials(os.Stdin)
+	if err != nil {
+		return ExitError(ctx, ExitUnsupported, err, "Error: %v while parsing git-credential", err)
+	}
+	// try git/host/username... If username is empty, simply try git/host
+	path := "git/" + fsutil.CleanFilename(cred.Host) + "/" + fsutil.CleanFilename(cred.Username)
+	if !s.Store.Exists(ctx, path) {
+		// if the looked up path is a directory with only one entry (e.g. one user per host), take the subentry instead
+		if s.Store.IsDir(ctx, path) {
+			tree, err := s.Store.Tree(ctx)
+			if err != nil {
+				return ExitError(ctx, ExitDecrypt, err, "Error: %v while listing the storage", err)
+			}
+			sub, err := tree.FindFolder(path)
+			if err != nil {
+				return ExitError(ctx, ExitDecrypt, err, "Error: %v while listing the storage", err)
+			}
+			entries := sub.List(0)
+			if len(entries) == 1 {
```

Comment:
Negate and return early:

```go
if len(entries) != 1 {
	fmt.Fprintln(os.Stderr, "gopass error: too many entries")
}
path = "git/" + entries[0]
```
Not a major issue in this case, but would prefer it that way.